lib/lightstep/transport/http_json.rb

```diff
@@ -25,28 +26,43 @@ class HTTPJSON < Base
       # @param access_token [String] access token for LightStep server
       # @return [HTTPJSON]
       def initialize(host: LIGHTSTEP_HOST, port: LIGHTSTEP_PORT, verbose: 0, encryption: ENCRYPTION_TLS, access_token:)
-        @host = host
-        @port = port
         @verbose = verbose
-        @encryption = encryption
 
         raise Tracer::ConfigurationError, "access_token must be a string" unless String === access_token
         raise Tracer::ConfigurationError, "access_token cannot be blank"  if access_token.empty?
         @access_token = access_token
+
+        # This mutex protects the use of our Net::HTTP instance which we
+        # maintain as a long lived connection. While a Lightstep::Transport is
+        # typically called only from within the reporting thread, there are
+        # some situations where this can be bypassed (directly calling `flush`
+        # for example)
+        @mutex = Mutex.new
+
+        @http = Net::HTTP.new(host, port)
+        @http.use_ssl = encryption == ENCRYPTION_TLS
+        @http.keep_alive_timeout = 5
       end
 
       # Queue a report for sending
       def report(report)
         p report if @verbose >= 3
 
-        https = Net::HTTP.new(@host, @port)
-        https.use_ssl = @encryption == ENCRYPTION_TLS
         req = Net::HTTP::Post.new('/api/v0/reports')
         req['LightStep-Access-Token'] = @access_token
         req['Content-Type'] = 'application/json'
         req['Connection'] = 'keep-alive'
+
         req.body = report.to_json
-        res = https.request(req)
+
+        @mutex.synchronize do
+          # Typically, keep-alive for Net:HTTP is handled inside a start block,
+          # but that's awkward with our threading model. By starting it manually,
+          # once, the TCP connection should remain open for multiple report calls.
+          @http.start unless @http.started?
```

Comment:
Since finish is never called, will this leave dangling connections open to the satellite endpoint when the process shuts down?

What happens when the server sends a `Connection:close` header back with this class? Will an exception be raised, will http.started become false?